compiler-core/src/metadata.rs

```diff
@@ -1,10 +1,71 @@
 //! Seriaisation and deserialisation of Gleam compiler metadata into binary files
-//! using the Cap'n Proto schema.
-
-mod module_decoder;
-mod module_encoder;
+//! using the Serde and Bincode
 
 #[cfg(test)]
 mod tests;
 
-pub use self::{module_decoder::ModuleDecoder, module_encoder::ModuleEncoder};
+use crate::type_::{generic_var, Type, TypeVar};
+use crate::uid::UniqueIdGenerator;
+use crate::{type_::Module, Result};
+use std::sync::Arc;
+
+pub fn encode(data: &Module) -> Result<Vec<u8>> {
+    let span = tracing::info_span!("metadata");
+    let _enter = span.enter();
+
+    let buffer = bincode::serialize(data).expect("failed to encode metadata (bincode)");
+
+    Ok(buffer)
+}
+
+fn undo_links(id_generator: &UniqueIdGenerator, type_var: TypeVar) -> Arc<Type> {
+    match type_var {
+        TypeVar::Link { type_ } => regenerate_type_ids(id_generator, &type_),
+        TypeVar::Generic { .. } => generic_var(id_generator.next()),
+        TypeVar::Unbound { .. } => panic!("unexpected `TypeVar::Unbound` in cache decoding"),
+    }
+}
+
+fn regenerate_type_ids(id_generator: &UniqueIdGenerator, type_: &Type) -> Arc<Type> {
+    match (*type_).clone() {
+        Type::App {
+            public,
+            module,
+            name,
+            args,
+        } => Arc::new(Type::App {
+            public,
+            module,
+            name,
+            args: args
+                .into_iter()
+                .map(|arg| regenerate_type_ids(&id_generator, &arg))
+                .collect(),
+        }),
+        Type::Fn { args, retrn } => Arc::new(Type::Fn {
+            args: args
+                .into_iter()
+                .map(|arg| regenerate_type_ids(&id_generator, &arg))
+                .collect(),
+            retrn: regenerate_type_ids(&id_generator, &retrn),
+        }),
+        Type::Var { type_ } => undo_links(id_generator, (*type_).clone().into_inner()),
+        Type::Tuple { elems } => Arc::new(Type::Tuple {
+            elems: elems
+                .into_iter()
+                .map(|elm| regenerate_type_ids(&id_generator, &elm))
+                .collect(),
+        }),
+    }
+}
```

Comment:
Could you use the type constructor functions I linked before for these please?

Reply:
done for all except `App` as I couldn't fine one